pkg/vmconfig/crypto/crypto_reconciler_pre.go

```diff
@@ -571,17 +578,108 @@ func onRecrypt(
 		},
 	}
 
+	recryptedDisks := onRecryptDisks(args)
+
 	logger.Info(
 		"Recrypt VM",
 		"currentKeyID", args.curKey.id,
 		"currentProviderID", args.curKey.provider,
 		"newKeyID", args.newKey.id,
 		"newProviderID", args.newKey.provider,
-		"newProviderIsDefault", args.newKey.isDefaultProvider)
+		"newProviderIsDefault", args.newKey.isDefaultProvider,
+		"recryptedDisks", recryptedDisks)
 
 	return 0, nil, nil
 }
 
+func onRecryptDisks(args reconcileArgs) []string {
+	var fileNames []string
+	for i := range args.moVM.Config.Hardware.Device {
+		baseDev := args.moVM.Config.Hardware.Device[i]
+		if disk, ok := baseDev.(*vimtypes.VirtualDisk); ok {
+			if disk.VDiskId == nil { // Skip FCDs
+
+				switch tBack := disk.Backing.(type) {
+				case *vimtypes.VirtualDiskFlatVer2BackingInfo:
+					if tBack.KeyId != nil {
+						if updateDiskBackingForRecrypt(args, disk) {
+							fileNames = append(fileNames, tBack.FileName)
+						}
+					}
+				case *vimtypes.VirtualDiskSeSparseBackingInfo:
+					if tBack.KeyId != nil {
+						if updateDiskBackingForRecrypt(args, disk) {
+							fileNames = append(fileNames, tBack.FileName)
+						}
+					}
+				case *vimtypes.VirtualDiskSparseVer2BackingInfo:
+					if tBack.KeyId != nil {
+						if updateDiskBackingForRecrypt(args, disk) {
+							fileNames = append(fileNames, tBack.FileName)
+						}
+					}
+				}
+			}
+		}
+	}
+	return fileNames
+}
+
+func updateDiskBackingForRecrypt(
+	args reconcileArgs,
+	disk *vimtypes.VirtualDisk) bool {
+
+	devSpec := getOrCreateDeviceChangeForDisk(args, disk)
+	if devSpec == nil {
+		return false
+	}
+
+	if devSpec.Backing == nil {
+		devSpec.Backing = &vimtypes.VirtualDeviceConfigSpecBackingSpec{}
+	}
+
+	// Set the device change's crypto spec to be the same as the VM's.
+	devSpec.Backing.Crypto = args.configSpec.Crypto
+
+	return true
+}
+
+// getOrCreateDeviceChangeForDisk returns the device change for the specified
+// disk. If there is an existing Edit change, that is returned. If there is an
+// existing Add/Remove change, nil is returned. Otherwise a new Edit change is
+// returned.
+func getOrCreateDeviceChangeForDisk(
+	args reconcileArgs,
+	disk *vimtypes.VirtualDisk) *vimtypes.VirtualDeviceConfigSpec {
+
+	for i := range args.configSpec.DeviceChange {
```

Comment:
nit: Some of these should never be nil, right? If they happen to be, seems it would be better to panic than silently ignore? Easier to read without too imho.

```go
	for _, dc := range args.configSpec.DeviceChange {
		ds := dc.GetVirtualDeviceConfigSpec()
		vd := ds.Device.GetVirtualDevice()
		if vd.Key == disk.Key {
			if ds.Operation == vimtypes.VirtualDeviceConfigSpecOperationEdit {
				return ds
			}
			return nil
		}
	}
```

Reply:
- `args.configSpec.DeviceChange[]` - our code should not be adding a nil to this list
- `dc.GetVirtualDeviceConfigSpec()` - in this case just converts `vimtypes.BaseVirtualDeviceConfigSpec` to `*vimtypes.VirtualDeviceConfigSpec`
- `ds.Device` - can't have a DeviceChange w/o this field. If this is nil, it's a bug in our code
- `ds.Device.GetVirtualDevice()`- returns a pointer to a field value, can't be nil if we've made it this far